c/driver_manager/adbc_driver_manager.cc

```diff
@@ -138,6 +320,176 @@ struct ManagedLibrary {
     // release() from the DLL - how to handle this?
   }
 
+  AdbcStatusCode GetDriverInfo(const std::string_view driver_name,
+                               const uint16_t load_options, DriverInfo& info,
+                               struct AdbcError* error) {
+    if (driver_name.empty()) {
+      SetError(error, "Driver name is empty");
+      return ADBC_STATUS_INVALID_STATE;
+    }
+
+    std::filesystem::path driver_path(driver_name);
+    const bool allow_relative_paths =
+        load_options & static_cast<uint16_t>(kAdbcAllowRelativePaths);
+    if (driver_path.has_extension()) {
+      if (driver_path.is_relative() && !allow_relative_paths) {
+        SetError(error, "Driver path is relative and relative paths are not allowed");
+        return ADBC_STATUS_INVALID_STATE;
+      }
+
+      // if the extension is .toml, attempt to load the manifest
+      // erroring if we fail
+#ifdef _WIN32
+      if (strncasecmp(driver_path.extension().string().c_str(), ".toml", 5) == 0) {
+#else
+      if (driver_path.extension() == ".toml") {
+#endif
+        auto status = LoadDriverManifest(driver_path, info, error);
+        if (status == ADBC_STATUS_OK) {
+          return Load(info.lib_path.c_str(), error);
+        }
+        return status;
+      }
+
+      // if the extension is not .toml, then just try to load the provided
+      // path as if it was an absolute path to a driver library
+      return Load(driver_path.string().c_str(), error);
+    }
+
+    if (driver_path.is_absolute()) {
+      // if we have an absolute path without an extension, first see if there's a
+      // toml file with the same name.
+      driver_path.replace_extension(".toml");
+      if (std::filesystem::exists(driver_path)) {
+        auto status = LoadDriverManifest(driver_path, info, error);
+        if (status == ADBC_STATUS_OK) {
+          return Load(info.lib_path.c_str(), error);
+        }
+      }
+
+      // otherwise just try to load the provided path as if it was an absolute path
+      return Load(driver_name.data(), error);
+    }
+
+    if (driver_path.has_extension()) {
+      if (driver_path.is_relative() &&
+          (load_options & static_cast<uint16_t>(kAdbcAllowRelativePaths)) == 0) {
+        SetError(error, "Driver path is relative and relative paths are not allowed");
+        return ADBC_STATUS_INVALID_STATE;
+      }
+
+#if defined(_WIN32)
+      static const std::string kPlatformLibrarySuffix = ".dll";
+      if (strncasecmp(driver_path.extension().string().c_str(),
+                      kPlatformLibrarySuffix.c_str(),
+                      kPlatformLibrarySuffix.size()) == 0) {
```

Comment:
`.DLLX` may be matched too.

It may be better that we have a helper function that compares a path extension.